File: zone-server/src/main/kotlin/net/bestia/zone/trade/TradeStateSMSG.kt
package net.bestia.zone.trade

import net.bestia.bnet.proto.EnvelopeProto
import net.bestia.bnet.proto.TradeStateSmsgProto
import net.bestia.zone.item.container.ReservedItem
import net.bestia.zone.message.SMSG
import net.bestia.zone.util.EntityId

/**
 * The whole trade as one recipient sees it - which is why "own" and "partner" mean different things in the
 * two copies sent for the same trade.
 *
 * Always a full snapshot. A refused action is answered by re-sending this, which snaps an optimistic client
 * back without either end needing a per-refusal code.
 */
data class TradeStateSMSG(
  val tradeId: Long,
  val status: Status,
  val partnerMasterName: String,
  val partnerEntityId: EntityId,
  val ownOffer: List<ReservedItem>,
  val partnerOffer: List<ReservedItem>,
  val ownLocked: Boolean,
  val partnerLocked: Boolean,
  val ownConfirmed: Boolean,
  val partnerConfirmed: Boolean,
) : SMSG {

  /** The subset of [TradeStatus] a client is ever shown; the rest are moments with no window to draw. */
  enum class Status { OPEN, LOCKED, COMPLETED, CANCELLED }

  override fun toBnetEnvelope(): EnvelopeProto.Envelope {
    val state = TradeStateSmsgProto.TradeStateSMSG.newBuilder()
      .setTradeId(tradeId)
      .setStatus(mapStatus(status))
      .setPartnerMasterName(partnerMasterName)
      .setPartnerEntityId(partnerEntityId)
      .addAllOwnOffer(ownOffer.map { mapItem(it) })
      .addAllPartnerOffer(partnerOffer.map { mapItem(it) })
      .setOwnLocked(ownLocked)
      .setPartnerLocked(partnerLocked)
      .setOwnConfirmed(ownConfirmed)
      .setPartnerConfirmed(partnerConfirmed)
      .build()

    return EnvelopeProto.Envelope.newBuilder()
      .setTradeState(state)
      .build()
  }

  private fun mapStatus(status: Status): TradeStateSmsgProto.TradeStatus = when (status) {
    Status.OPEN -> TradeStateSmsgProto.TradeStatus.TRADE_STATUS_OPEN
    Status.LOCKED -> TradeStateSmsgProto.TradeStatus.TRADE_STATUS_LOCKED
    Status.COMPLETED -> TradeStateSmsgProto.TradeStatus.TRADE_STATUS_COMPLETED
    Status.CANCELLED -> TradeStateSmsgProto.TradeStatus.TRADE_STATUS_CANCELLED
  }

  private fun mapItem(item: ReservedItem): TradeStateSmsgProto.TradeOfferItem =
    TradeStateSmsgProto.TradeOfferItem.newBuilder()
      .setOfferSlotId(item.offerSlotId)
      .setItemId(item.itemId.toInt())
      .setUniqueId(item.uniqueId)
      .setAmount(item.amount)
      .setDurability(item.durability)
      .setMaxDurability(item.maxDurability)
      .setSlots(item.slots)
      .setUpgradeLevel(item.upgradeLevel)
      .build()
}

File: bestia-client/src/Bnet/Message/Trade/TradeStateSMSG.cs
using Godot;

namespace BestiaBehemothClient.Bnet.Message.Trade
{
  /// <summary>
  /// How far a trade has got. Ordinals match <c>bnet.TradeStatus</c>.
  /// </summary>
  public enum TradeStatus
  {
    Open = 0,
    Locked = 1,
    Completed = 2,
    Cancelled = 3
  }

  /// <summary>
  /// The whole trade, as this client sees it. One message opens the window, updates it and closes it.
  ///
  /// Always a full snapshot, so the window can re-render both columns wholesale on every one. That is also
  /// how a refused action corrects itself: the server re-sends this and the optimistic client snaps back.
  /// </summary>
  [GlobalClass]
  public partial class TradeStateSMSG : ISMSG
  {
    [Export] public ulong TradeId { get; set; }

    [Export] public TradeStatus Status { get; set; } = TradeStatus.Open;

    /// <summary>
    /// <see cref="Status"/> as a lowercase string, for the GDScript side - it cannot see a C# enum's
    /// members, so matching on this avoids re-declaring the ordinals over there by hand.
    /// </summary>
    [Export] public string StatusName { get; set; } = "open";

    [Export] public string PartnerMasterName { get; set; } = string.Empty;

    [Export] public ulong PartnerEntityId { get; set; }

    [Export] public Godot.Collections.Array<TradeOfferItem> OwnOffer { get; set; } = [];

    [Export] public Godot.Collections.Array<TradeOfferItem> PartnerOffer { get; set; } = [];

    /// <summary>Reserved for currency, which does not exist yet - always 0.</summary>
    [Export] public ulong OwnGold { get; set; }

    [Export] public ulong PartnerGold { get; set; }

    [Export] public bool OwnLocked { get; set; }

    [Export] public bool PartnerLocked { get; set; }

    [Export] public bool OwnConfirmed { get; set; }

    [Export] public bool PartnerConfirmed { get; set; }

    public static TradeStateSMSG FromProto(global::Bnet.TradeStateSMSG proto)
    {
      var own = new Godot.Collections.Array<TradeOfferItem>();
      foreach (var item in proto.OwnOffer)
      {
        own.Add(TradeOfferItem.FromProto(item));
      }

      var partner = new Godot.Collections.Array<TradeOfferItem>();
      foreach (var item in proto.PartnerOffer)
      {
        partner.Add(TradeOfferItem.FromProto(item));
      }

      var status = MapStatus(proto.Status);

      return new TradeStateSMSG
      {
        TradeId = proto.TradeId,
        Status = status,
        StatusName = status.ToString().ToLowerInvariant(),
        PartnerMasterName = proto.PartnerMasterName,
        PartnerEntityId = proto.PartnerEntityId,
        OwnOffer = own,
        PartnerOffer = partner,
        OwnGold = proto.OwnGold,
        PartnerGold = proto.PartnerGold,
        OwnLocked = proto.OwnLocked,
        PartnerLocked = proto.PartnerLocked,
        OwnConfirmed = proto.OwnConfirmed,
        PartnerConfirmed = proto.PartnerConfirmed
      };
    }

    private static TradeStatus MapStatus(global::Bnet.TradeStatus status)
    {
      return status switch
      {
        global::Bnet.TradeStatus.Locked => TradeStatus.Locked,
        global::Bnet.TradeStatus.Completed => TradeStatus.Completed,
        global::Bnet.TradeStatus.Cancelled => TradeStatus.Cancelled,
        _ => TradeStatus.Open
      };
    }
  }
}
